src/network/networkd-netlabel.c

```diff
@@ -0,0 +1,197 @@
+         * operations.
+         */
+        masked_addr = address->in_addr;
+        r = in_addr_mask(address->family, &masked_addr, address->prefixlen);
+        if (r < 0)
+                return r;
+
+        if (address->family == AF_INET) {
+                r = sd_netlink_message_append_in_addr(m, NLBL_UNLABEL_A_IPV4ADDR, &masked_addr.in);
+                if (r < 0)
+                        return r;
+
+                r = sd_netlink_message_append_in_addr(m, NLBL_UNLABEL_A_IPV4MASK, &netmask.in);
+        } else if (address->family == AF_INET6) {
+                r = sd_netlink_message_append_in6_addr(m, NLBL_UNLABEL_A_IPV6ADDR, &masked_addr.in6);
+                if (r < 0)
+                        return r;
+
+                r = sd_netlink_message_append_in6_addr(m, NLBL_UNLABEL_A_IPV6MASK, &netmask.in6);
+        }
+        if (r < 0)
+                return r;
+
+        r = netlink_call_async(address->link->manager->genl, NULL, m, netlabel_handler, link_netlink_destroy_callback,
+                               address->link);
+        if (r < 0)
+                return r;
+
+        link_ref(address->link);
+        return 0;
+}
+
+static const char *get_label(const Address *address) {
+        assert(address);
+        assert(address->link);
+
+        if (!address->link->network || !IN_SET(address->family, AF_INET, AF_INET6))
+                return NULL;
+
+        switch (address->source) {
+        case NETWORK_CONFIG_SOURCE_DHCP4:
+                return address->link->network->dhcp_netlabel;
+        case NETWORK_CONFIG_SOURCE_DHCP6:
+                return address->link->network->dhcp6_netlabel;
+        case NETWORK_CONFIG_SOURCE_DHCP_PD:
+                return address->link->network->dhcp_pd_netlabel;
+        case NETWORK_CONFIG_SOURCE_NDISC:
+                return address->link->network->ndisc_netlabel;
+        case NETWORK_CONFIG_SOURCE_STATIC:
+                return address->netlabel;
+        default:
+                return NULL;
+        }
+}
+
+void address_add_netlabel(const Address *address) {
+        int r;
+        assert(address);
+
+        const char *label;
+        label = get_label(address);
+        if (!label)
+                return;
+
+        r = netlabel_command(NLBL_UNLABEL_C_STATICADD, label, address);
+        if (r < 0)
+                log_link_warning_errno(address->link, r, "Adding NetLabel %s for IP address %s failed, ignoring", label,
+                                       IN_ADDR_PREFIX_TO_STRING(address->family, &address->in_addr, address->prefixlen));
+        else
+                log_link_debug(address->link, "Adding NetLabel %s for IP address %s", label,
+                               IN_ADDR_PREFIX_TO_STRING(address->family, &address->in_addr, address->prefixlen));
+}
+
+void address_del_netlabel(const Address *address) {
+        int r;
+        assert(address);
+
+        const char *label;
+        label = get_label(address);
+        if (!label)
+                return;
```

Comment:
Is this necessary? On removal, the netlink message does not contain label. I think we should unconditionally remove netlabel for the address.

Reply:
That could conflict with labels configured with for example `netlabelctl` when there's no networkd configuration for NetLabels. I'd only remove the label if and only if we installed it.